### DISCUSSION TITLE
How do you estimate the ideal airbox size?

### DISCUSSION THREAD

**Discussion by KJ7LNW** [thliebig/openEMS-Project#159] [Q&A Usage] [answered]
This question (probably) has different answers depending on the type of geometry.  For the purpose of this question, I am curious about these two use cases:
 - Antenna far-fields
 - Microstrip filters

Antenna questions:
- Does the air box need to be a multiple of the wavelength? 
- Does the air box space _around or beyond_ the model need to be a multiple of the wavelength?
- Does the airbox minimum axis size change with polarization? For example, does a vertically polarized antenna only need additional airbox space in the "vertical" axis for that wave?

Microstrip questions:
- Same questions as above if relevant
- Under what conditions does the model touch the air box? I noticed this wall touching configuration in @toammann's "OpenEMS Simulation Model" section of (#157), and was curious when it is appropriate for the model to touch the wall, and what considerations are involved in such a case.

 Thank you for your help!

**Comment by biergaizi** [comment, marked as answer]
> Does the air box need to be a multiple of the wavelength?
>
> Does the air box space around or beyond the model need to be a multiple of the wavelength?

If PML is used, the distance between the radiating structure and PML should be *at least* lambda/4.

Otherwise, no.

> Does the airbox minimum axis size change with polarization? For example, does a vertically polarized antenna only need additional airbox space in the "vertical" axis for that wave?

No, I don't think it changes with polarization. Instead, it changes with the radiation pattern. If there's little or no near-field and far-field around a region in the box, there's no need to ensure the clearance between the structure and PML, or even using a PML at all.

------

It's not a really a problem of "air box size", but a problem of selecting the proper boundary conditions, and leaving an appropriate space between your simulated structure and the PML absorbing boundary condition to stay clear from fringe field. 

1. **Boundary Conditions**: In FDTD, electromagnetic field only exists in a small simulation box, so the question is what to do when the waves reach the edge. Mathematically it's known as the boundary condition of partial differential equations.

2. **PEC and PMC**:  The simplest solution is setting the electric or magnet field outside the box to zero, mathematically it's known as the Dirichlet boundary condition, physically it represents the behavior of a Perfect Electric Conductor or a Perfect Magnetic Conductor. In analogues to short or open circuit in 1D transmission lines, they reflect the waves back into the box. PEC is useful when you're actually simulating a lossless shielded box or ground plane, such as a cavity resonator, a metal enclosure, the bottom layer of a two-layer circuit board, or a waveguide. In this case, it's okay to touch the simulation box's boundary. You also do not need to model any metal at boundary if you decide to use the PEC shortcut, since the PEC boundary conditions already implicitly represent a conductor. It's also worth noting that PEC has the fastest simulation speed since it does not require extra calculations.

3. **ABC**: For open boundary problems, the reflection must be eliminated - this includes most transmission line and antenna simulations - there must be an Absorbing Boundary Condition (ABC) to make the waves disappear at the boundary, similar to a termination resistor in 1D.

    3.1. **Mur's ABC**: The first generation of ABCs are differential-equation based, such as Liao's ABC or Mur's ABC. Mur's ABC is implemented in openEMS. It's unconditionally stable, the computational overhead is low, and can be successfully used if used with care. However, the absorption is imperfect - complete absorption occurs if and only if the angle of incidence is normal to the ABC. Unwanted reflections from imperfect absorption can create significant simulation errors.

    3.2. **PML**: The next breakthrough was the invention of the *Perfectly Matched Layer*. PML is not a true mathematical boundary condition, but behaves as an unusual artificial material that can magically absorb almost all the E&M waves perfectly. Thus, PML is easier to use and can reduce reflections much better than Mur's ABC. There are many mathematical formulations of PML, the one used in openEMS is called Uniaxial PML (UPML). 

4. **PML Precautions**: Several precautions should be followed when using PML.

    4.1. **Avoid Touching PML Cells**:  PML is a material, not a true mathematical boundary condition, so it occupies physical space in the simulation box. Using `PML_8`, it means the PML material occupies the last 8 cells at edge of the mesh grid, and one must manually ensure that simulated structure does not overlap with the last 8 cells of a surface at which PML is turned on - even though PML is not visualized in the 3D model. In some cases, a thicker layer of this virtual "EMI shielding foam" is needed, so while `PML_8` is a good default, it can also be changed between 6 to 20 cells.

    4.2. **Stay Clear from Evanescent Waves**: UPML can effectively absorb radiative waves, but it cannot absorb evanescent waves (i.e. near field, fringe field). In fact, UPML can become numerically unstable if fringe fields are entering it - the "perfectly" matched layer is not so "perfect" after all. Thus, one must ensure that there's a proper distance between the radiating structure, at least a distance of lambda/4. *Interestingly, research literature says another PML formulation known as Convolutional PML (CPML) works better than UPML under evanescent waves, supporting it in openEMS may be a topic of future discussion.*

5. **Computational Overhead**: PEC and PMC have the lowest computing overhead, Mur's ABC has a modest overhead. PML has the highest overhead due to separate update passes on the last 8 cells near the boundary, not just at the boundary. In fact, openEMS's UPML implementation is not optimized and incurs an even higher overhead. Thus, sometimes it's desirable to minimize the use of PML if one needs high performance. A simulation box has 6 surfaces, It can be a win to avoid the use of PML boundary conditions at least on *some* surfaces - but note that it's a tricky optimization. In general  it's better to start with an all-PML box first and finish debugging the simulation until it works before applying further optimizations, which should be compared with the all-PML result as a sanity check.

    5.1. **Using Different Boundary Conditions for Speed**: For example, if you're modeling a planar antenna on a printed circuit board, in a high-performance model, you can probably omit the bottom metal surface and use PEC instead. If the antenna is radiating towards the right side of the box, the rightmost surface is set to PML. Other surfaces which do not experience a strong incidental electromagnetic field can be set to Mur's ABC (or PML with a smaller separation distance if you know there's little fringe field here). 

Since my focus is programming and I haven't done many practical simulations, I can't give any practical advise, the above is an explanation of the general principle. I'll leave the rest to other experts like @toammann.

**Reply by thliebig** [reply]
@biergaizi You always write so nice summaries about different topic. It would be really great if you could contribute that all to the openEMS documentation?
I guess the first topic we should add is how to build the doc locally to see what changes look like?

**Reply by biergaizi** [reply]
I actually have an unfinished draft called "Elements of FDTD", which attempts to explain five basic concepts of FDTD simulations to beginners, including (1) 3D modeling, (2) meshing, (3) boundary conditions, (4) ports and excitation, and (5) post-processing. But I found it's rather difficult to write a comprehensive review, and because later I started working on engine optimization rather than documentation, it was never finished...

**Reply by KJ7LNW** [reply]
> and because later I started working on engine optimization rather than documentation, it was never finished...

@biergaizi, please publish what you have, any documentation in FOSS projects helps, even if its "incomplete"

+1 for the answer above to be in the docs, too.

**Reply by nico-arnold** [reply]
@biergaizi Thanks for the good write-up, I learned a lot again and especially about why some of my earlier simulations made problems. (: 
That draft you're working on would indeed be really helpful, looking forward to that one!
